hpy/devel/src/runtime/argparse.c

```diff
@@ -34,54 +100,108 @@ int _HPyArg_ParseItem(HPyContext ctx, HPy current_arg, const char **fmt, va_list
         break;
     }
     case 'O': {
-        HPy *output = va_arg(*vl, HPy *);
-        _BREAK_IF_OPTIONAL(current_arg);
-        *output = current_arg;
+        if (**fmt == '+') {
+            (*fmt)++;
+            HPy *output = va_arg(*vl, HPy *);
+            _BREAK_IF_OPTIONAL(current_arg);
+            *output = HPy_Dup(ctx, current_arg);
+            if (HPy_IsNull(*output))
+                return 0;
+            break;
+        }
+        else {
+            HPy *output = va_arg(*vl, HPy *);
+            _BREAK_IF_OPTIONAL(current_arg);
+            *output = current_arg;
+        }
         break;
     }
     default:
-        HPyErr_SetString(ctx, ctx->h_ValueError, "XXX: Unknown arg format code");
+        HPyErr_SetString(ctx, ctx->h_SystemError, "XXX: Unknown arg format code");
         return 0;
     }
     return 1;
 }
 
+static const char *
+parse_errstring(const char *fmt, const char **err_funcname, const char **err_message)
+{
+    const char *fmt1 = fmt;
+
+    for (fmt1 = fmt; *fmt1 != 0; fmt1++) {
+        if (*fmt1 == ':') {
+            *err_funcname = fmt1 + 1;
+            break;
+        }
+        if (*fmt1 == ';') {
+            *err_message = fmt1 + 1;
+            break;
+        }
+    }
+    return fmt1;
+}
+
+static void
+set_error(HPyContext ctx, HPy exc, const char *err_funcname, const char *err_message, const char *msg) {
+    char err_buf[_MAX_ERR_STRING_LENGTH];
+    if (err_message != NULL) {
+        snprintf(err_buf, _MAX_ERR_STRING_LENGTH, "%s", err_message);
+    }
+    if (err_message == NULL) {
```

Comment:
```suggestion
    else {
```

I'm not sure if this is the right logic but it might be. I haven't really followed through the code.

Reply:
Thanks! This was leftover from refactoring `set_error` a few times.